crosslink issue on linux, and cleanup entire dir

structwafel opened the pull request:
I had issues when installing extensions, due to `Invalid cross-device link (os error 18)`

```
[2025-02-21][02:25:42][ERROR] Fail to install tarball (vscode): Error: failed to rename old path: /tmp/0f8fab02-ae32-4889-9b5a-28fc622df2ef/package to new path: /home/lgx/.local/share/sh.kunkun.desktop/extensions/vscode with error: Invalid cross-device link (os error 18)
```

Apparently on linux at times the tmp can be a different file system?
https://stackoverflow.com/questions/42392600/oserror-errno-18-invalid-cross-device-link

So need to manually copy each file instead of renaming the path. sad :(



Also before it would keep the `/tmp/uuidv4/` destinations, as the `/tmp/uuidv4/package` is the `decompressDest`.

structwafel committed:
crosslink issue on linux, and cleanup entire dir

HuakunShen commented:
You may have extensions and tmp on different drives.
Thanks.
But I think if a `copy_dir_all` is required, it should be implemented in Rust for better performance. 
There should be existing crate for this in rust.

structwafel committed:
moved copy_dir_all to rust

structwafel committed:
using dircpy instead of diy due to complications

structwafel commented:
@HuakunShen 
Have moved the functionality to rust. added dircpy as a dependency:
https://github.com/woelper/dircpy/
https://crates.io/crates/dircpy

Seems like there is more than copying files (like copying metadata etc).


I think this is needed in general, as arch linux uses tmpfs for the `/tmp` dir (https://wiki.archlinux.org/title/Tmpfs)
We can check if the `/tmp` dir is in the same filesysytem ect, but unsure if it is worth it for this.
Copying and deleting is fast enough for this operation?

HuakunShen commented:
Extensions are small, generally 14kb to 1MB. I don't think copying and deleting will have any performance issue.
I may also need to expose this fs API to extensions.

HuakunShen commented:
I will need to move this command to the Jarvis plugin for extension to access. I will do it very soon.

HuakunShen committed:
refactor: move copy_dir_all to jarvis plugin

All commands are in jarvis plugin, this is more organized.
And this API will be exposed to extensions.
